### PR TITLE
[Codegen] Support dynamic/scalable sizes when folding insert_slice into xfer_write

### PR DESCRIPTION
**Pull request by MacDue**
This enables further optimizations which are currently missed when targeting scalable vectors.

### CODE REVIEW

**Comment by MacDue** [comment]
~I'm unable to reproduce the crash from https://github.com/iree-org/iree/actions/runs/10010028276/job/27671399332?pr=17963.  I'd be really helpful if the CI had `llvm-symbolizer` setup, I can't see what could crash here...~

Gone after rebase.

File: compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp [hunk 1]
```diff
@@ -117,6 +119,47 @@ class FoldExtractSliceIntoTransferRead final
...
+
+  auto dest = writeOp.getSource();
+  auto destShape = writeOp.getShapedType().getShape();
+  auto destDimSize =
+      [&](unsigned dimIndex) -> FailureOr<iree_compiler::DimBoundSize> {
+    auto size = destShape[dimIndex];
+    if (size != ShapedType::kDynamic)
+      return iree_compiler::DimBoundSize{size};
+    return iree_compiler::computeDimUpperBound(dest, dimIndex, vscaleRange);
+  };
```
Comment (banach-space) [inline compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp]: This is a basically a function - please follow the naming convention for functions:
* https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly

Unless there's some other guideline for lambdas?

Also, could you document what this lambda does? Specifically for those of us unfamiliar with `computeDimUpperBound`?

File: compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp [hunk 2]
```diff
@@ -117,6 +119,47 @@ class FoldExtractSliceIntoTransferRead final
   }
 };
 
+/// Returns true if `writeOp` fully overwrites its destination.
```
Comment (banach-space) [inline compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp:122]: Please add an example.

File: compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp [hunk 3]
```diff
@@ -117,6 +119,47 @@ class FoldExtractSliceIntoTransferRead final
...
+
+  std::optional<iree_compiler::VscaleRange> vscaleRange;
+  auto vecType = writeOp.getVectorType();
+  if (vecType.isScalable()) {
+    auto targetAttr =
+        iree_compiler::IREE::HAL::ExecutableTargetAttr::lookup(writeOp);
+    vscaleRange = iree_compiler::getDefaultVscaleRange(targetAttr);
+  }
+
+  auto dest = writeOp.getSource();
```
Comment (banach-space) [inline compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp]: [nit] spell out auto

File: compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp [hunk 4]
```diff
@@ -117,6 +119,47 @@ class FoldExtractSliceIntoTransferRead final
...
+  auto destShape = writeOp.getShapedType().getShape();
```
Comment (banach-space) [inline compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp]: [nit] spell out auto

File: compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp [hunk 5]
```diff
@@ -117,6 +119,47 @@ class FoldExtractSliceIntoTransferRead final
...
+  auto destDimSize =
+      [&](unsigned dimIndex) -> FailureOr<iree_compiler::DimBoundSize> {
+    auto size = destShape[dimIndex];
+    if (size != ShapedType::kDynamic)
+      return iree_compiler::DimBoundSize{size};
+    return iree_compiler::computeDimUpperBound(dest, dimIndex, vscaleRange);
+  };
+
+  auto vecShape = vecType.getShape();
+  auto vecScalableFlags = vecType.getScalableDims();
```
Comment (banach-space) [inline compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp]: [nit] spell out auto

File: compiler/src/iree/compiler/Codegen/Common/test/optimize_tensor_insert_extract_slices.mlir
Comment (banach-space) [inline compiler/src/iree/compiler/Codegen/Common/test/optimize_tensor_insert_extract_slices.mlir:1]: Missing `CHECK-NOT: tensor.insert_slice`

**Review by banach-space — COMMENTED** [review]
Nice!

IIUC, this is meant to extend `isDestinationFullyOverwritten`? That and nothing else?

EDIT:
Sorry, I meant `FoldInsertSliceIntoTransferWrite` rather than `isDestinationFullyOverwritten` 🤦🏻

File: compiler/src/iree/compiler/Codegen/Common/test/optimize_tensor_insert_extract_slices.mlir
Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/Common/test/optimize_tensor_insert_extract_slices.mlir:1]: I've just checked the whole functions via `CHECK-NEXT` now (as it's only two lines).

File: compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp [hunk 6]
```diff
@@ -117,6 +119,63 @@ class FoldExtractSliceIntoTransferRead final
...
+/// Returns true if `writeOp` fully overwrites its destination.
+///
+/// Example:
+///
+/// ```MLIR
+/// vector.transfer_write %vec, %dest[%c0, %c0] {in_bounds = [true, true]}
+///    : vector<4x5xf32>, tensor<4x5xf32>
+/// ```
+///
+/// This is an easy case `vector<4x5xf32>` fully-overwrites `tensor<4x5xf32>` as
```
Comment (banach-space) [inline compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp]: [nit]
```suggestion
/// This is an easy case, `vector<4x5xf32>` fully-overwrites `tensor<4x5xf32>` as
```

**Review by banach-space — COMMENTED** [review]
LG, but please add a negative tests as well.

**Review by banach-space — APPROVED** [review]
LGTM, thanks!

@hanhanW , would you like to take a look or shall we land it?

**Comment by hanhanW** [comment]
> LGTM, thanks!
> 
> @hanhanW , would you like to take a look or shall we land it?

I'll take a look today, sorry for the late review.

File: compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp [hunk 7]
```diff
@@ -117,6 +119,63 @@ class FoldExtractSliceIntoTransferRead final
   }
 };
 
+/// Returns true if `writeOp` fully overwrites its destination.
+///
+/// Example:
+///
+/// ```MLIR
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp]: nit: let's drop `MLIR` for consistency in this file. (side note: IMO it is a visual noise..)

```suggestion
/// ```
```

File: compiler/src/iree/compiler/Codegen/Common/test/optimize_tensor_insert_extract_slices.mlir [hunk 8]
```diff
@@ -63,6 +63,116 @@ func.func @fold_extract_slice_consumer_into_xfer_write_3(%arg0: vector<1x64x128x
...
+  %0 = vector.transfer_write %v, %extract_slice[%c0, %c0] {in_bounds = [true, true]} : vector<4x[2]xf32>, tensor<4x?xf32>
+  %1 = tensor.insert_slice %0 into %t2[%a, %b] [4, %c4_vscale] [1, 1] : tensor<4x?xf32> into tensor<?x?xf32>
+  return %1 : tensor<?x?xf32>
+}
+// CHECK-LABEL: func.func @negative_fold_insert_slice_into_transfer_write_scalable
+// CHECK: %[[WRITE:.*]] = vector.transfer_write
+// CHECK: tensor.insert_slice %[[WRITE]]
+
+
+// -----
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/test/optimize_tensor_insert_extract_slices.mlir]: nit: remove extra blank line.

```suggestion

// -----
```

File: compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp [hunk 9]
```diff
@@ -117,6 +119,63 @@ class FoldExtractSliceIntoTransferRead final
...
+/// dynamic tensors, where it resolves the tensor sizes via value-bounds
+/// analysis, and then checks if the vector type fully overwrites the tensor.
+static bool isDestinationFullyOverwritten(vector::TransferWriteOp writeOp) {
+  if (writeOp.hasOutOfBoundsDim())
+    return false;
+  if (writeOp.getVectorType().getRank() != writeOp.getShapedType().getRank())
+    return false;
+  if (writeOp.getMask())
+    return false;
+
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp:142]: Should we also check if the values in the `in_bounds` attribute are all `true` here? I think it is not fully overwritten case if one of them is false. Do I miss something?

**Review by hanhanW — CHANGES_REQUESTED** [review]
My concern was that there is a TODO saying that the pattern should be deprecated. Then I start thinking "when do we want the folding". My answer is "after hoisting redundant vector transfers", which is what the pass is doing. The TODO was added long time ago, and I guess we don't know the answer at the moment. The extract_slice ops and insert_slice ops  are really just the artifacts of tiling, so it is okay to fold them away after vectorization (in IREE). Thus my concern is gone. The change looks okay to me. Just one question and few nits.

```
// TODO: this is brittle and should be deprecated in favor of a more general
// pattern that applies on-demand.
```

File: compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp (hunk 9 above)
Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp:142]: That's checked on the first `if` (of `isDestinationFullyOverwritten()`):
```c++
 if (writeOp.hasOutOfBoundsDim())
    return false;
```

**Comment by MacDue** [comment]
~~New test failures come from #17804, need to wait for #18113.~~

File: compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp (hunk 9 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp:142]: ah, I missed, thanks!

**Review by hanhanW — APPROVED** [review]

**Comment by hanhanW** [comment]
The failing test seems to be flaky, re-running it now. I'll land it once CI is happy.

**Pull request merged**